### PR TITLE
[1.1.11] Optimize UDF function logic

### PR DESCRIPTION
**Pull request by huangKai-2323**
### What is the purpose of the change

Optimize UDF function logic：
1. UDF User Management Permission Control
2. UDF classification management permission control and bug handling

### Related issues/PRs

Related issues: #

### Brief change log

- Define the core abstraction and interfaces of the EngineConn Factory;
- Define the core abstraction and interfaces of Executor Manager.

### Checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [ ] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [Linkis mailing list](https://linkis.apache.org/community/how-to-subscribe) first)
- [ ] **If this is a code change**: I have written unit tests to fully verify the new behavior.

### CODE REVIEW

**Comment by casionone** [comment]
Need add new files about UdfBaseInfoMapper ?

File: linkis-dist/package/db/upgrade/1.4.0_schema/mysql/linkis_ddl.sql [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+ALTER TABLE `linkis_ps_udf_manager`  ADD COLUMN `create_time` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP COMMENT 'create time';
+ALTER TABLE `linkis_ps_udf_manager`  ADD COLUMN `update_time` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP COMMENT 'update_time';
+
+ALTER TABLE `linkis_ps_udf_shared_group`  ADD COLUMN `create_time` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP COMMENT 'create time';
+ALTER TABLE `linkis_ps_udf_shared_group`  ADD COLUMN `update_time` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP COMMENT 'update_time';
```
Comment (casionone) [inline linkis-dist/package/db/upgrade/1.4.0_schema/mysql/linkis_ddl.sql:22]: These fields need to be added when the data is updated 
UDFDao.xml

File: linkis-dist/package/db/upgrade/1.4.0_schema/mysql/linkis_ddl.sql [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+ALTER TABLE `linkis_ps_udf_shared_info`  ADD COLUMN `create_time` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP COMMENT 'create time';
+ALTER TABLE `linkis_ps_udf_shared_info`  ADD COLUMN `update_time` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP COMMENT 'update_time';
+
+ALTER TABLE `linkis_ps_udf_user_load`  ADD COLUMN `create_time` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP COMMENT 'create time';
+ALTER TABLE `linkis_ps_udf_user_load`  ADD COLUMN `update_time` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP COMMENT 'update_time';
+
+
+ALTER TABLE `linkis_ps_udf_version`  ADD COLUMN `update_time` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP COMMENT 'update_time' AFTER `create_time`;
```
Comment (casionone) [inline linkis-dist/package/db/upgrade/1.4.0_schema/mysql/linkis_ddl.sql:31]: no need AFTER `create_time`?

File: linkis-public-enhancements/linkis-basedata-manager/src/main/java/org/apache/linkis/basedatamanager/server/restful/UdfTreeRestfulApi.java [hunk 3]
```diff
@@ -64,10 +73,25 @@ public Message list(
...
     ModuleUserUtils.getOperationUser(
         request, "Query all data of UDF Tree,search name:" + searchName);
-    List<UdfTreeEntity> udfTreeEntityList = udfTreeService.list();
+    List<UdfTreeEntity> udfTreeEntityList = new ArrayList<>();
+    UdfTreeEntity entity = new UdfTreeEntity();
+    entity.setCategory(category);
+    QueryWrapper<UdfTreeEntity> querySysWrapper =
+        new QueryWrapper<>(entity)
+            .eq("category", entity.getCategory())
+            .in("user_name", "sys", "expire", "share", "bdp");
```
Comment (casionone) [inline linkis-public-enhancements/linkis-basedata-manager/src/main/java/org/apache/linkis/basedatamanager/server/restful/UdfTreeRestfulApi.java:85]: adjust as configuration item 
key is UDF_FUN_SYSTEM_CATEGORY
linkis.udf.fun.system.category="user_name,sys,expire,share,bdp"

File: linkis-public-enhancements/linkis-basedata-manager/src/main/java/org/apache/linkis/basedatamanager/server/restful/UdfTreeRestfulApi.java [hunk 4]
```diff
@@ -100,9 +124,27 @@ public Message add(HttpServletRequest request, @RequestBody UdfTreeEntity udfTre
...
-    boolean result = udfTreeService.removeById(id);
-    return Message.ok("").data("result", result);
+    String username =
+        ModuleUserUtils.getOperationUser(request, "Remove a UDF Tree Record,id:" + id.toString());
+    if (!Configuration.isAdmin(username)) {
+      return Message.error("User '" + username + "' is not admin user[非管理员用户]");
+    }
+    UdfTreeEntity entity = udfTreeService.getById(id);
+    if (null != entity && entity.getParent() == -1) {
+      return Message.error("Root level directory deletion prohibited！");
```
Comment (casionone) [inline linkis-public-enhancements/linkis-basedata-manager/src/main/java/org/apache/linkis/basedatamanager/server/restful/UdfTreeRestfulApi.java:134]: The root directory is forbidden to delete["根目录禁止删除"]

File: linkis-public-enhancements/linkis-basedata-manager/src/main/java/org/apache/linkis/basedatamanager/server/restful/UdfTreeRestfulApi.java [hunk 5]
```diff
@@ -100,9 +124,27 @@ public Message add(HttpServletRequest request, @RequestBody UdfTreeEntity udfTre
...
+        new QueryWrapper<>(new UdfTreeEntity()).eq("parent", id);
+    List<UdfTreeEntity> folderList = udfTreeService.list(queryWrapper);
+    QueryWrapper<UdfBaseInfoEntity> udfQueryWrapper =
+        new QueryWrapper<>(new UdfBaseInfoEntity()).eq("tree_id", id);
+    List<UdfBaseInfoEntity> functoinList = udfBaseinfoService.list(udfQueryWrapper);
+    if (CollectionUtils.isEmpty(folderList) && CollectionUtils.isEmpty(functoinList)) {
+      boolean result = udfTreeService.removeById(id);
+      return Message.ok("").data("result", result);
+    } else {
+      return Message.error("Please manually delete the sub folder or function and try again!");
```
Comment (casionone) [inline linkis-public-enhancements/linkis-basedata-manager/src/main/java/org/apache/linkis/basedatamanager/server/restful/UdfTreeRestfulApi.java:146]: Please delete the subdirectory first[请先删除子目录]

**Comment by casionone** [comment]
```
Warning: arning] Using a password on the command line interface can be insecure.
Warning: arning] Using a password on the command line interface can be insecure.
ERROR 1064 (42000) at line 208 in file: './linkis-dist/package/db/linkis_ddl.sql': You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '(`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8' at line 7
Error: Process completed with exit code 1.
```
sql check failed! 
https://github.com/WeDataSphere/linkis/actions/runs/4784815554/jobs/8506770978?pr=131

File: linkis-dist/package/db/upgrade/1.4.0_schema/mysql/linkis_ddl.sql (hunk 2 above)
Comment (huangKai-2323) [reply linkis-dist/package/db/upgrade/1.4.0_schema/mysql/linkis_ddl.sql:31]: done

File: linkis-public-enhancements/linkis-basedata-manager/src/main/java/org/apache/linkis/basedatamanager/server/restful/UdfTreeRestfulApi.java [hunk 6]
```diff
@@ -64,10 +73,25 @@ public Message list(
 
   @ApiOperation(value = "all", notes = "Query all data of UDF Tree", httpMethod = "GET")
   @RequestMapping(path = "/all", method = RequestMethod.GET)
-  public Message all(HttpServletRequest request, String searchName) {
+  public Message all(HttpServletRequest request, String searchName, String category) {
```
[the remark this answers is not in the document]
Comment (huangKai-2323) [reply linkis-public-enhancements/linkis-basedata-manager/src/main/java/org/apache/linkis/basedatamanager/server/restful/UdfTreeRestfulApi.java:76]: done

File: linkis-public-enhancements/linkis-basedata-manager/src/main/java/org/apache/linkis/basedatamanager/server/restful/UdfTreeRestfulApi.java (hunk 3 above)
Comment (huangKai-2323) [reply linkis-public-enhancements/linkis-basedata-manager/src/main/java/org/apache/linkis/basedatamanager/server/restful/UdfTreeRestfulApi.java:85]: done

File: linkis-public-enhancements/linkis-basedata-manager/src/main/java/org/apache/linkis/basedatamanager/server/restful/UdfTreeRestfulApi.java (hunk 4 above)
Comment (huangKai-2323) [reply linkis-public-enhancements/linkis-basedata-manager/src/main/java/org/apache/linkis/basedatamanager/server/restful/UdfTreeRestfulApi.java:134]: done

File: linkis-public-enhancements/linkis-basedata-manager/src/main/java/org/apache/linkis/basedatamanager/server/restful/UdfTreeRestfulApi.java (hunk 5 above)
Comment (huangKai-2323) [reply linkis-public-enhancements/linkis-basedata-manager/src/main/java/org/apache/linkis/basedatamanager/server/restful/UdfTreeRestfulApi.java:146]: done

**Pull request merged**